### PR TITLE
use ament_cmake_export_interfaces

### PR DESCRIPTION
**Pull request by dirk-thomas**
Exports the created library target using `ament_cmake_export_interfaces`.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
* [Build Status](http://ci.ros2.org/job/ci_linux/2359/)
* [Build Status](http://ci.ros2.org/job/ci_osx/1813/)
* [Build Status](http://ci.ros2.org/job/ci_windows/2470/)

**Comment by dirk-thomas** [comment]
Waiting for review.

**Comment by wjwwood** [comment]
What is the purpose of the pr? There's no context here.

**Comment by dirk-thomas** [comment]
The purpose is to export the library target (instead / beside exporting the include directory and library path). Then the downstream package can use it as an imported target (see referenced PR). This is the "new" way in CMake compared to the "old" way.

File: ament_index_cpp/CMakeLists.txt [hunk 1]
```diff
@@ -14,9 +14,12 @@ add_library(${PROJECT_NAME} SHARED
   src/get_search_paths.cpp
   src/has_resource.cpp)
 target_compile_definitions(${PROJECT_NAME} PRIVATE "AMENT_INDEX_CPP_BUILDING_DLL")
-target_include_directories(${PROJECT_NAME} PUBLIC include)
+target_include_directories(${PROJECT_NAME} PUBLIC
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+  "$<INSTALL_INTERFACE:include>")
```
Comment (wjwwood) [inline ament_index_cpp/CMakeLists.txt:19]: Why is this necessary? Does `$<INSTALL_INTERFACE:include>` even exist at build time of this target?

File: ament_index_cpp/CMakeLists.txt [hunk 2]
```diff
@@ -38,6 +41,7 @@ install(DIRECTORY include/
   DESTINATION include)
 
 install(TARGETS ${PROJECT_NAME}
+  EXPORT export_${PROJECT_NAME}
```
Comment (wjwwood) [inline ament_index_cpp/CMakeLists.txt:44]: Is this necessary, given the code in the `ament_export_interfaces` code?

https://github.com/ament/ament_cmake/blob/master/ament_cmake_export_interfaces/cmake/ament_cmake_export_interfaces_package_hook.cmake#L28-L33

File: ament_index_cpp/CMakeLists.txt [hunk 3]
```diff
@@ -14,9 +14,12 @@ add_library(${PROJECT_NAME} SHARED
   src/get_search_paths.cpp
   src/has_resource.cpp)
 target_compile_definitions(${PROJECT_NAME} PRIVATE "AMENT_INDEX_CPP_BUILDING_DLL")
-target_include_directories(${PROJECT_NAME} PUBLIC include)
+target_include_directories(${PROJECT_NAME} PUBLIC
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+  "$<INSTALL_INTERFACE:include>")
 
 ament_export_include_directories(include)
+ament_export_interfaces(export_${PROJECT_NAME})
```
Comment (wjwwood) [inline ament_index_cpp/CMakeLists.txt:22]: Even after reading the documentation for `ament_export_interfaces()` I can't tell, is this an input variable or an output variable? Is `export_${PROJECT_NAME}` set before this is called?

**Review by wjwwood — APPROVED** [review]
I can't really tell you if the code is fine, as I've never used most of these features. But I don't recognize anything wrong with the changes.

I still don't get the motivation, you've clarified what the code is doing, but I don't know why you're doing it or why you're just doing it to this target.

Is there something you're unable to do right now that this allows? Are you just trying a new feature out? I have no idea.

File: ament_index_cpp/CMakeLists.txt (hunk 1 above)
Comment (dirk-thomas) [reply ament_index_cpp/CMakeLists.txt:19]: The `BUILD_INTERFACE` is being used to build the library. The `INSTALL_INTERFACE` is being used when downstream packages try to link against this target. See https://cmake.org/cmake/help/v3.5/command/target_include_directories.html

File: ament_index_cpp/CMakeLists.txt (hunk 3 above)
Comment (dirk-thomas) [reply ament_index_cpp/CMakeLists.txt:22]: The argument has to match an `EXPORT` of an `install(..)` call. See comment below.

File: ament_index_cpp/CMakeLists.txt (hunk 2 above)
Comment (dirk-thomas) [reply ament_index_cpp/CMakeLists.txt:44]: This install rule is being "identified" by the name `export_${PROJECT_NAME}`. `ament_export_interfaces(<same_identifier>)` creates the necessary part for the CMake config file that the installed targets can be imported in downstream packages.

**Comment by dirk-thomas** [comment]
Using imported targets is a cleaner approach to pass information along to downstream packages. Compared to the classic approach of using `*_INCLUDE_DIRS`, `*-LIBRARIES` a downstream package can now:

* link against a specific library if an upstream package provides multiple
* link against the imported target without worrying what is "attached" to it (include dirs, transitive libraries, transitive include dirs, definition, compile flags, anything)
* information are not expanded to absolute paths until they are being used in a downstream package

While `ament` supported this for a long time we have never used it until now. Imo existing code should in the future move to this pattern instead of the classic approach to leverage these advantages.

**Comment by wjwwood** [comment]
> While `ament` supported this for a long time we have never used it until now. Imo existing code should in the future move to this pattern instead of the classic approach to leverage these advantages.

I understand and agree it's a cool feature and a better way to expose information, but what you still haven't answered, imo, is why now and why this target in this package rather than all packages or N packages or something else? The motivation seems to be "because".

**Comment by dirk-thomas** [comment]
> I understand and agree it's a cool feature and a better way to expose information, but what you still haven't answered, imo, is why now and why this target in this package rather than all packages or N packages or something else? The motivation seems to be "because".

I try to squeeze in time for reducing technical debt whenever I can. It just happened to be the case that I picked this item (exporting / importing targets). I picked this package since it has exactly one downstream dependency and was therefore easy to convert.

**Comment by wjwwood** [comment]
> I try to squeeze in time for reducing technical debt whenever I can. It just happened to be the case that I picked this item (exporting / importing targets). I picked this package since it has exactly one downstream dependency and was therefore easy to convert.

Ok, that's fine, but in the future, if you want people to review it quickly, I'd recommend you add some context. I hesitated at the time you opened them because I didn't know if it was related to the console logging work or some other issue you were debugging and I didn't know if more was to come. If it's just trying it out or burning down some technical debt, just say so.

**Comment by dirk-thomas** [comment]
> I hesitated at the time you opened them because I didn't know if it was related to the console logging work or some other issue you were debugging and I didn't know if more was to come.

I understand the "in review" label that it means that a PR is ready to be reviewed and mergable. If that is not yet the case the ticket explicitly mentions something like "just for feedback" or "please review but more is coming".

**Comment by wjwwood** [comment]
I'm just telling you why I didn't review it immediately, which was there was no context or motivation so I was uncertain as to its purpose or priority. I guess others felt the same way as no one else reviewed it. You could avoid this in the future by being more generous with details from the onset. It's just advice you don't have to follow it.
